### ISSUE TITLE
Parsing of Netex zip file provided by Austria fails with "UicOperatingPeriod cannot be cast to class OperatingPeriod"

### ISSUE THREAD

**Issue by twatzl** [entur/netex-parser-java#90]
Hello,

I am trying to parse NETEX provided at https://data.mobilitaetsverbuende.at/de/data-sets

Fahrplandaten Eisenbahn (NeTEx) 2023 gesamt

But I get the following error:

```
[main] INFO org.entur.netex.loader.parser.ResourceFrameParser - Netex import - Element skipped: javax.xml.bind.JAXBElement
[main] INFO org.entur.netex.loader.parser.ResourceFrameParser - Netex import - Element skipped: javax.xml.bind.JAXBElement
[main] INFO org.entur.netex.loader.parser.ResourceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.OperationalContextsInFrame_RelStructure
[main] INFO org.entur.netex.loader.parser.ResourceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ResponsibilitySetsInFrame_RelStructure
[main] INFO org.entur.netex.loader.parser.ResourceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.VehicleTypesInFrame_RelStructure
[main] INFO org.entur.netex.loader.parser.ResourceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.TypeOfFrameRefStructure
[main] INFO org.entur.netex.loader.parser.NetexParser - Netex import - Element skipped: org.rutebanken.netex.model.TypeOfFrameRefStructure
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.ServiceJourneyPattern
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.TransfersInFrame_RelStructure
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.DirectionsInFrame_RelStructure
[main] INFO org.entur.netex.loader.parser.ServiceFrameParser - Netex import - Element skipped: org.rutebanken.netex.model.TypeOfFrameRefStructure
Exception in thread "main" java.lang.ClassCastException: class org.rutebanken.netex.model.UicOperatingPeriod cannot be cast to class org.rutebanken.netex.model.OperatingPeriod (org.rutebanken.netex.model.UicOperatingPeriod and org.rutebanken.netex.model.OperatingPeriod are in unnamed module of loader 'app')
	at org.entur.netex.loader.parser.ServiceCalendarFrameParser.parseOperatingPeriods(ServiceCalendarFrameParser.java:93)
	at org.entur.netex.loader.parser.ServiceCalendarFrameParser.parse(ServiceCalendarFrameParser.java:40)
	at org.entur.netex.loader.parser.ServiceCalendarFrameParser.parse(ServiceCalendarFrameParser.java:27)
	at org.entur.netex.loader.parser.NetexDocumentParser.parse(NetexDocumentParser.java:108)
	at org.entur.netex.loader.parser.NetexDocumentParser.parseCommonFrame(NetexDocumentParser.java:63)
	at org.entur.netex.loader.parser.NetexDocumentParser.parseCompositeFrame(NetexDocumentParser.java:101)
	at org.entur.netex.loader.parser.NetexDocumentParser.parseCommonFrame(NetexDocumentParser.java:83)
	at org.entur.netex.loader.parser.NetexDocumentParser.parseFrameList(NetexDocumentParser.java:53)
	at org.entur.netex.loader.parser.NetexDocumentParser.parse(NetexDocumentParser.java:48)
	at org.entur.netex.loader.parser.NetexDocumentParser.parseAndPopulateIndex(NetexDocumentParser.java:42)
	at org.entur.netex.NetexParser.load(NetexParser.java:87)
	at org.entur.netex.NetexParser.parse(NetexParser.java:51)
	at org.entur.netex.NetexParser.parse(NetexParser.java:32)
	at MainKt.main(Main.kt:6)

Process finished with exit code 1
```

main.kt:

```
import org.entur.netex.NetexParser

fun main(args: Array<String>) {
    val parser = NetexParser()
    val index = parser.parse("path/to/file.zip")


    println(index.stopPlaceIndex.latestVersions.size);
}
```

build.gradle.kts:

```
plugins {
    kotlin("jvm") version "1.8.0"
    application
}

group = "io.opendata"
version = "1.0-SNAPSHOT"

repositories {
    mavenCentral()
}

dependencies {
    implementation("org.entur:netex-parser-java:2.0.1")
    implementation("org.slf4j:slf4j-simple:2.0.7")
    testImplementation(kotlin("test"))
}

tasks.test {
    useJUnitPlatform()
}

kotlin {
    jvmToolchain(11)
}

application {
    mainClass.set("MainKt")
}
```

I am using Java 20. 
There is no info about compatible versions, required additional dependencies or a minimum working example in README.md. 

What could be the issue? Maybe you could provide some more information on how to use the library.

Best Tobi

**Comment by twatzl** [comment]
I just tried it with the zip file provided in the /src/test/resources of this repo and the code works fine. So it seems to be some issue particularly connected to the data I am trying to parse. 

Any support with debugging this woudl be highly appreciated though. :)

**Comment by testower** [comment]
Hi and thanks for reporting this. There is probably limited support for parsing data that is not following the Nordic NeTEx profile, but I'll try and dig into this. The issue seems to that the parser tries to cast the type `UicOperatingPeriod` (which is not part of the Nordic profile) to `OperatingPeriod`.

The difference between `UicOperatingPeriod` and `OperatingPeriod` seems to be that UicOperating period has the following additional properties: `ValidDayBits` and `DaysOfWeek`.

Your proposed solution to change the type to `OperatingPeriod_VersionStructure` could possibly work in the sense that it compiles, but any data set on those additional properties will be lost, so not sure if that's what you want.

**Comment by twatzl** [comment]
I see. Sorry for late response. I just now have time again to dig into this. 

Another possible solution could be to use UicOperatingPeriod as a default and just set valid day bits to all 1 and DaysOfWeek to all. 

Unfortunately the official resources from EU regarding Netex are woefully undermaintained. Some reference websites are even no longer reachable as the domains have not been renewed. So as far as I have seen this is the only parsing library available for Java.

I will probably have to dig further into this as I am trying to parse NeTeX data from all over Europe. Maybe then I can come up with a solution. But don't expect it anytime soon. I am following this in my free time whenever I have some to spare.

**Comment by testower** [comment]
> Another possible solution could be to use UicOperatingPeriod as a default and just set valid day bits to all 1 and DaysOfWeek to all.

I don't think it's a good idea to map between different types in a parsing library like this. The parsed entities should reflect what's actually in the XML.

I think the best we can do is to allow an additional index for the UicOperatingPeriod. The library user will have to know which index to check (or check both).

**Comment by skinkie** [comment]
I think what you are using is a very specific parser towards the Norwegian Profile?

Below is what I am doing with https://github.com/entur/netex-java-model which works on _any_ file.

```java
        try {
            InputStream inputStreamStream = new FileInputStream(fileName);

            JAXBElement<PublicationDeliveryStructure> publicationDeliveryStructureJAXBElement = (JAXBElement<PublicationDeliveryStructure>) jaxbUnmarshaller.unmarshal(inputStreamStream);
            return publicationDeliveryStructureJAXBElement.getValue();
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }
```

I am currently moving my years of investment in Java NeTEx code towards Python in concert with xsData. In my perspective this gives much more flexibility.

**Comment by testower** [comment]
@skinkie Indeed it works on any file. The purpose of *this* library is to index entities from the NeTEx files for easy lookup. Indeed it has been developed in the context of the *Nordic* profile, as mentioned previously, but I don't see any problem with expanding it to support parts of NeTEx not used by us. Just like netex-java-model.

**Comment by skinkie** [comment]
@testower if only the indexing was implemented _correctly_ I would have agreed, sadly it isn't. https://github.com/entur/netex-parser-java/blob/master/src/main/java/org/entur/netex/index/impl/VersionedNetexEntityIndexImpl.java#L65

As said, I have many years of experience in doing this _correctly_ in Java. Which includes the id + version attributes and resolving of version=any. Given my experience building code on top of these kind of shortcuts is like building on quicksand. You can't use these shortcuts if you are parsing profiles that trusts on the correct implementation, and for example provides an object of the same type with the same id, but a different version. Sure you can get away with it in the Nordic and EPIP context, now try VDV462 or the Dutch profile.

**Comment by testower** [comment]
Again, I have never claimed this library is built for anything other than the Nordic profile, and it serves us well. It is mostly based on the NeTEx parser in OTP, so yes, it has a very specific purpose for us.

I don't doubt your experience @skinkie but this is now clearly off-topic. If you want to contribute on what you think is a weakness in the library, I welcome you to open an issue to describe it beyond point to a line in the code with a claim that it is incorrect.

**Comment by skinkie** [comment]
@testower I have been asked by a collegue of you to give some hints to @twatzl. Going forward with this one on EPIP is not something I would recommend.

And I understand it _could_ work on the Nordic profile.

**Comment by testower** [comment]
I have exactly 0 experience with anything other than the Nordic profile, so @twatzl I would definitely follow @skinkie's advise!

As for the versioned entities index, yes it is profile-specifc: Versions must be increasing positive integers. Explicitly non-versioned entities (version=any) don't end up in this index. Not in our datasets anyway.

**Comment by twatzl** [comment]
Alright. Thanks for the advice so far. 
The reason I came here is actually that this is the only netex parsing library I could find for Java. Unfortunately the standard page of European Union refers to some "sample implementations", which no longer exist. So you rather recommend parsing plain XML for the most compatibility around Europe? Or are there any other parsing libraries?

As I said I am just a random dev from Austria trying to write some kind of route planner/parse netex data into a graph database to get an overview of what data is available across Europe and what is missing. :D So unfortunately I also don't have a lot of knowledge regarding abbreviations or how the system is designed (specifically profiles and what EPIP is).

You mentioned OTP. Just had to look that up as well. Seems they had a similar issue. https://github.com/opentripplanner/OpenTripPlanner/issues/3640

And holy heck do I read that correctly that the standard on an EU level is not yet agreed upon/ there is no reference profile? What kind of rabbit hole did I get myself into here? 😅🙈

Thanks for all the answers and the your time so far!

**Comment by skinkie** [comment]
> Alright. Thanks for the advice so far. The reason I came here is actually that this is the only netex parsing library I could find for Java. Unfortunately the standard page of European Union refers to some "sample implementations", which no longer exist. So you rather recommend parsing plain XML for the most compatibility around Europe? Or are there any other parsing libraries?

It is critical to understand no "NeTEx implementation" implements it as full. Hence the only thing you get is (almost) for free is a _deserialiser_ into objects. Depending on the implementation (and practically speaking for which profile it was intended) allows you de read and interact with in. Transformations between one view into another, or automatic references which are possible for some simpeler identity-key-constraint based schema's, have **not** been defined.

The code I have been working on in Java is the transformation stuff, but still would require to set up (Hash)Map's to resolve the references. But it can transform virtually every variant as long and you would know what 'rules' apply to them. It can't blind parse an undefined profile.

For Python I am able to read an arbitrairy file into a relational database, retaining the object structure. Relationships behing here the first class citizen, but as might be visible from my latest pull requests in the NeTEx repository would require a far thighter definition that it has now. When Entur says all my objects carry a object based unique version number, while a German or Dutch standard defines that versions are groups of objects related to eachother then you know that you can store that information but require in the two variants a completely different interaction pattern.

> As I said I am just a random dev from Austria trying to write some kind of route planner/parse netex data into a graph database to get an overview of what data is available across Europe and what is missing. :D So unfortunately I also don't have a lot of knowledge regarding abbreviations or how the system is designed (specifically profiles and what EPIP is).

If you are just transforming EPIP my suggestion would be get out the ServiceJourneyPattern and ServiceJourney/timetabledPassingTimes. That allows you to build a graph. But I personally would not transform NeTEx into something different, if the something different _by itself_ is not the target.

> And holy heck do I read that correctly that the standard on an EU level is not yet agreed upon/ there is no reference profile? What kind of rabbit hole did I get myself into here? 😅🙈

The situation is quite different. There is a European standard consider that a _vocabulary_, which is actually based on another _vocabulary_ (Transmodel). The semantics within Transmodel are generally high level defined, so low level details miss, how objects relate to each other is preset. 

Because NeTEx is _huge_ subsets of the standard are defined as _profiles_. Here, sometimes with a custom XSD which restricts a bit more, is defined which elements are mandatory in that profile. Still validation should be possible with the full standard.

Norway defined their Nordic profile as one of the first parties. But that definition does all kind of tricks that no regular XML parser can handle. And if it can handle it (includes), nothing has been currently defined in the Nordic profile to allow a regular parser to open separate files, and run over it as a single instance. You need application logic in order to handle this.
 
In general can be said that the Nordic profile uses a partial denormalised timetable. Where the German (VDV462) or Dutch profiles are normalised, and the original Swiss effort was fully denormalised (semantically like GTFS is). France had historically something similar. Because none of the parties wanted to choose a proper direction (either fully normalised or fully denormalised) the European Passenger Information Profile (EPIP) is a half baked denormalised timetable which has ServiceJourneyPatterns and TimetabledPassingTimes while the ServiceJourneyPattern ab-so-lu-te-ly does not make any sense, one could just have used the "Calls" like happens in SIRI. Net effect: huge, but virtually impossible to interprete wrong files, but _obviously_ not compatble with the early Nordic profile.

So yes, there is a reference profile (EPIP), where virtually everything is pruned out, which is supposed to be available at the National Access Points. But as the history laid out above it is not compatible with the work Entur did. And to be fair to them, they invested a lot of time and man power, but kept being stubborn otherwise they would have switched to an extended EPIP profile by now making their efforts much more compatible.

**Comment by twatzl** [comment]
Thank you so much for all those insights. These will help me a lot to get started.

Closing the ticket now. As it's basically an unfixable issue.

**Comment by testower** [comment]
I don't agree that the original issue posted here is *unfixable*. It is definitely fixable by applying the solution I outlined, which is to add an index for UicOperatingPeriod. As for the rest of the discussion, I don't have anything to add, other than saying that specifically the versioning issue doesn't seem impossible to fix: #108
